src/GraphQL.Tests/Subscription/SampleObservable.cs

```diff
@@ -0,0 +1,80 @@
+#nullable enable
+
+namespace GraphQL.Tests.Subscription;
```

Comment:
All that code looks very similar (simplified, of course) to `Subject<T>` from `System.Reactive` package. `Subject<T>` is an observable and an observer - both at the same time. For the last couple of days I read the documentation and looked Into Rx source code. Did you write this file intentionally to eliminate `Subject<T>` usage? I still see reference to `System.Reactive.Linq` from tests so I think this file may be deleted.

Reply:
I am simply not familiar to the System.Reactive package.  I did try to use one of the classes (a 'Replay' class perhaps?), but it did not execute synchronously and so messed up my tests.  I would need to use some other code from System.Reactive to asynchronously wait for notifications, and then the tests would pass again.

I can take a look again.

Reply:
`ReplaySubject<T>` is another class inherited from `SubjectBase`, `Subject<T>` inherits `SubjectBase` too and simpler than  `ReplaySubject<T>`. In fact `Subject<T>` does nothing special - just provides methods from observable and observer - without buffering, scheduling, etc. It does just that you already wrote in `SampleObservable`.

Reply:
I tested it.  Everything passes except my tests that simulate what happens if data arrives subsequent to an `OnError` or `OnComplete` notification.  Unfortunately `Subject<T>` automatically unsubscribes when an error or completion is posted, and so I cannot use `Subject<T>` to test those scenarios.  I am going to leave the `SampleObservable` there so that all scenarios can be covered.

Now it is considered "best practice" according to MS for the observable to disconnect and not send any more events after the error is reported.  As such, the observer should disconnect itself after an error is received.  It is important to note that this behavior is at the discretion of the observable, and not something we can control.

As it relates to our code, I currently have OnError events converted into a GraphQL response and sent to the client with the standard GQL_DATA message.  The client is still responsible for closing the connection via GQL_STOP; if more data arrives, it will get processed normally.  It does not seem to fit the protocol to send a GQL_ERROR message in the middle of the stream; the error is only supposed to be in response to the initial connection attempt.  However, I could change the code to immediately close the connection after an error is posted, sending a GQL_COMPLETE message after the GQL_DATA packet.

Although it would seem most of those details pertain to the server project, I should probably make a change here so that after OnError wraps the error in a GraphQL response which is passed along via OnData (as it is now), there immediately follows an OnComplete notification.  Either that, or just wrap the exception in an ExecutionError and pass it along via OnError, for the server project to wrap in a GraphQL response and etc.  Any thoughts?

Again, options are:

1. Wrap exceptions from OnError into ExecutionResult but do not terminate observable sequence; let the connected client detect the error and terminate/retry connection, or
2. Wrap exceptions from OnError into ExecutionResult and pass along via OnData, followed by an immediate OnComplete notification forcing the client to retry the connection if desired,
3. Wrap exceptions from OnError as ExecutionError and pass along to GraphQL Server to handle (and decide from the above options).

Reply:
Right now the code is set up for "option 1".  I think I'll change it to "option 3".  This allows more flexibility based on the exact protocol in use.

Reply:
Eh, it's a lot of work to change all the tests.  I'll wait until tomorrow in case you have any comments.

Reply:
I find it difficult to give a full answer, I still do not know the subscriptions well. But I know that the MS in the documentation says that exception object from `OnError` method should be treated as an **informational** message. I guessed that the unsubscribing from observers in `Subject.OnError` could break your tests. Well, OK, let's keep `SampleObservable` as is.

Reply:
> However, I could change the code to immediately close the connection after an error is posted, sending a GQL_COMPLETE message after the GQL_DATA packet.

🤔 I would not if this error is a part of data message.

Reply:
I meant, if OnError is received from the observable source (which cannot be processed by the GraphQL schema), then and only then, wrap into an execution result, send it to the connected client, and send a disconnect packet. 

Any errors that happen during processing of the data packet (e.g. resolver errors) are just data packets and wouldn’t terminate the stream.

Reply:
Either way it seems that I should probably pass along the error to the server so the server can take the correct course of action rather than masking OnError events here

Reply:
Done.  This change will need a matching change in the server project along these lines for the same behavior as before:

```cs
public void OnError(Exception error)
{
    // pass along exceptions to the client
    if (error is ExecutionError executionError)
        OnNext(new ExecutionResult { Errors = new ExecutionErrors { executionError } });
    else
        OnNext(new ExecutionResult { Errors = new ExecutionErrors { new ExecutionError($"Unhandled error of type {error.GetType().Name}") } });
    // optional: disconnect client
    // OnComplete();
}
```

The only time there would be an exception that is not of type `ExecutionError` in the above sample is when the document executer's unhandled exception delegate throws an exception.

Reply:
See:
- https://github.com/graphql-dotnet/server/pull/754